programming-languages.md

```diff
@@ -1,51 +1,82 @@
+core software development tasks.
 
-These are (alphabetically): Java, Python, Ruby. These are successfully being
-used by teams at the moment, and we are confident in how to host and operate
-applications written in them. New development should normally be carried out
-in one of them.
+This should make it easier for developers to:
 
-There will be sensible reasons to go outside this core group. For example:
+- move around the organisation
+- develop shared components
+- improve their personal development
+- master how they operate applications
 
-- when we're extending an existing codebase/ecosystem
-- when scripting in a particular environment
-- when we're experimenting during an alpha (with an expectation that it's replaced by something from the core set for beta)
+## Java, Python and Ruby
 
-The set of languages in the core group will change over time. This is likely
-to be through the adoption of a new language into the core group, and with
-an existing language being put on hold for new work.
+We've chosen these 3 languages because they are successfully used by
+teams at the moment, and we are confident in how to host and operate
+applications written in them. We'd advise you to carry out new
+development in one of them.
 
-## Languages in use not in the core group
+## JavaScript
 
-### Go
+This recommendation only applies to server-side JavaScript.
+The service manual has information on
+[using client-side JavaScript][manual_js].
 
-A small amount of use in GOV.UK. Core language for CloudFoundry used by the
-PaaS team.
+[manual_js]: https://www.gov.uk/service-manual/technology/using-progressive-enhancement
 
-### Node.js (JavaScript)
+The only way we're comfortable writing server-side JavaScript at the
+moment is with [Node.js][nodejs].
 
-In use in GOV.UK Pay and Performance Platform. Not as mature as other
-languages/frameworks in use, and lacking a critical mass of skills in
-GDS, so further adoption discouraged. Note: does not apply to in-browser
-JavaScript.
+[nodejs]: https://nodejs.org/
 
-### Scala
+Node.js is used by a few projects at GDS; GOV.UK Frontend, GOV.UK Pay and the Performance Platform.
 
-Used in the early days of GDS. GOV.UK Licensing is the only remaining
-application written in Scala. Not for use in new development.
+In the past we've had problems operating and maintaining products written in
+Node.js. We suspect that this is because:
+
+- It's younger and less mature than Java, Python and Ruby
+- We have fewer developers with Node.js skills than other tools we use
+
+Because both of these points should gradually improve over time, we think
+that Node.js is suitable for creating thin, client-facing applications
+that don't persist data.
+
+## Go
+
+The GOV.UK router is built using Go, and it's the core language for Cloud
+Foundry which runs our Platform as a Service.
+
+We suspect that Go may be suitable for some systems programming, like proxying,
```

Comment:
My understanding from many discussions is that there is a general consensus of people that think Go is appropriate for use in systems programming, but a concern that we don't and won't have the skills in place to maintain these well when they're rarely used. How about the following to capture this:

> Go is an appropriate language for some systems programming, like proxying, routing and transforming HTTP requests.  However, since we don't maintain much Go and thus are likely to have few people with the skills to do so, any such code poses a significant risk.  Components should only be written in Go if there is no alternative maintained open source tool which could be used, and after a team has developed a plan to ensure that it will have the skills to maintain the component in future.